QT 5.9

scztt opened the pull request:
This branch upgrades SC to QT 5.8 / 5.9. The majority of the work was in incorporating the new WebEngine. 

Since it required more or less a complete rewrite, I added as much functionality as was easily exposed - thus the capabilities of WebView have been greatly expanded. The changes are documented in the help file - they are mostly backwards compatible, apart from 3 or so deprecated methods (marked as such).

This should be merged some time after the upcoming release - it's stable, but is definitely not a close-to-release sort of change.

brianlheim commented:
<3

IIRC there might still be issues with building this on Windows, related to a header include. I have win10 on a vm now (!) so I will test that out once i get a dev environment set up there

scztt commented:
Let me know - I'd be happy to triage anything weird/non-obvious if you can send me the VM too.

jamshark70 commented:
Ubuntu Studio 16.04 (current LTS), for instance, ships Qt 5.5.1 -- do we expect compatibility problems with that?

I'm not objecting, btw -- moving to WebEngine was absolutely necessary, kudos to scztt for that :+1:

scztt commented:
For whatever reason, Ubuntu distro's have always been radically behind on QT distributions. We've required a PPA-provided QT for the entire time we've been on the 5.x series, and probably much or all of the 4.x series as well. I wouldn't expect serious compatibility problems (I believe our travis build machines are using 16, for example...), since we've basically always been in this state. Qt 5.5 is 3 years old at this point, probably 4 by the time SC 3.10 is released with the new QT bump, so it doesn't seem unreasonably bleeding edge.

patrickdupuis commented:
>I wouldn't expect serious compatibility problems (I believe our travis build machines are using 16, for example...),

Are you certain that Travis CI is using Ubuntu 16.04? I remember @brianlheim recently saying that it uses 14.04.

Kudos for this PR indeed 👍

gusano commented:
> Are you certain that Travis CI is using Ubuntu 16.04

@patrickdupuis our builds use ubuntu-14 (trusty) because we [manually set it](https://github.com/supercollider/supercollider/blob/master/.travis.yml#L9). IMHO we should move on to ubuntu-16.
@scztt nice work! :+1:

jamshark70 commented:
> For whatever reason, Ubuntu distro's have always been radically behind on QT distributions. We've required a PPA-provided QT for the entire time we've been on the 5.x series, and probably much or all of the 4.x series as well. I wouldn't expect serious compatibility problems...

OK, on my machine, I'll deal with that when the time comes. I was just remembering a while back, when I had some weird GUI misbehavior in SC, and another user (Rainer, I think) suggested the problem was that I was using a Qt version/installation inappropriate for the OS. But if the recommendation will be to use a PPA, that's probably OK.

BTW there is no mention of any PPA in README_LINUX.md -- it's all news to me.

geoffroymontel commented:
Wow ! Thanks for that !

brianlheim commented:
@gusano

https://docs.travis-ci.com/user/reference/overview/

Where are you getting the info that 16.04 is available?

gusano commented:
@brianlheim it's not :) my mistake..

brianlheim commented:
@gusano np! i'd love to expand the build matrix, so when 16.04 becomes available i'd like to add it. thinking about the same with osx - we could potentially be testing against 3 versions of macOS. in fact it looks like they already added xcode9 even though it's still in beta!

brianlheim commented:
I'm just saying this because it hasn't been explicitly stated yet and I don't want anyone to be confused reading this thread – QtWebEngine is available since 5.4+.

Source: https://wiki.qt.io/New_Features_in_Qt_5.4

It looks like the travis 14.04 machines are using 5.8.

Not saying that there couldn't be compatibility problems, just wanted to clarify this single point.

gusano commented:
@brianlheim we should consider using [travis build stages](https://docs.travis-ci.com/user/build-stages/) as well then at some point.

brianlheim commented:
i've had a really hard time understanding how we would make use of that, tbh. it looks like maybe we could share a cache in parallel somehow, but i'm not sure how much time that would save since all those binaries would need to be copied either way.

telephon commented:
@scztt 
>This should be merged some time after the upcoming release - it's stable, but is definitely not a close-to-release sort of change.

is it ready to be merged into developer?

brianlheim commented:
@telephon this is completely untested on Windows right now, so let's please wait until someone has time to look at that.

telephon commented:
@passivist, you'd like to take a look?

orbsmiv commented:
FYI I've just merged this into the current _develop_ branch (along with _topic/menu support_) and compiled for macOS using Qt 5.10.0 and the build appears to be running without issue, although I haven't kicked its tyres much.
Is this still stalled due to the lack of Windows testing?

brianlheim commented:
@orbsmiv Thanks for the report! It was stalled because of the push to release 3.9, actually, but we will definitely pick this up soon. IIRC there's a compilation issue with MSVC that isn't too hard to fix, just a missing include somewhere.

brianlheim commented:
Just leaving a note that with 80aeac8 and 32908cd, this now builds with VS 2015 & Qt 5.9. Can't do 2017 yet because Boost 1.63 doesn't support that version of the compiler, or at least it doesn't recognize it so there are lots of config warnings. I think with some real-world testing we could merge this.

brianlheim commented:
@scztt - testing this out now. I get the following output when starting sclang built from this branch with Qt 5.10:

```
compiling class library (debug build)...
objc[97482]: Class QMacAutoReleasePoolTracker is implemented in both /Users/brianheim/git/supercollider/build_qt510/Install/SuperCollider/SuperCollider.app/Contents/Frameworks/QtCore.framework/Versions/5/QtCore (0x111b422c8) and /usr/local/Cellar/qt/5.10.0/lib/QtCore.framework/Versions/5/QtCore (0x11916f2c8). One of the two will be used. Which one is undefined.
objc[97482]: Class QT_ROOT_LEVEL_POOL__THESE_OBJECTS_WILL_BE_RELEASED_WHEN_QAPP_GOES_OUT_OF_SCOPE is implemented in both /Users/brianheim/git/supercollider/build_qt510/Install/SuperCollider/SuperCollider.app/Contents/Frameworks/QtCore.framework/Versions/5/QtCore (0x111b42340) and /usr/local/Cellar/qt/5.10.0/lib/QtCore.framework/Versions/5/QtCore (0x11916f340). One of the two will be used. Which one is undefined.
objc[97482]: Class RunLoopModeTracker is implemented in both /Users/brianheim/git/supercollider/build_qt510/Install/SuperCollider/SuperCollider.app/Contents/Frameworks/QtCore.framework/Versions/5/QtCore (0x111b42368) and /usr/local/Cellar/qt/5.10.0/lib/QtCore.framework/Versions/5/QtCore (0x11916f368). One of the two will be used. Which one is undefined.
QObject::moveToThread: Current thread (0x7fe84cd000a0) is not the object's thread (0x7fe84d8674e0).
Cannot move to target thread (0x7fe84cd000a0)

You might be loading two sets of Qt binaries into the same process. Check that all plugins are compiled against the right Qt binaries. Export DYLD_PRINT_LIBRARIES=1 and check that only one set of binaries are being loaded.
QObject::moveToThread: Current thread (0x7fe84cd000a0) is not the object's thread (0x7fe84d8674e0).
Cannot move to target thread (0x7fe84cd000a0)

You might be loading two sets of Qt binaries into the same process. Check that all plugins are compiled against the right Qt binaries. Export DYLD_PRINT_LIBRARIES=1 and check that only one set of binaries are being loaded.
	Found 715 primitives.
	Compiling directory '/Users/brianheim/git/supercollider/build_qt510/Install/SuperCollider/SuperCollider.app/Contents/Resources/SCClassLibrary'
WARNING: Could not open directory: '/Library/Application Support/SuperCollider/Extensions'
	To resolve this, either create the directory or remove it from your compilation paths.

	Compiling directory '/Users/brianheim/Library/Application Support/SuperCollider/Extensions'
Open ended string started on line 3 of file '/Users/brianheim/Library/Application Support/SuperCollider/Extensions/test.sc'
	Compiling directory '/Users/brianheim/Library/Application Support/SuperCollider/downloaded-quarks/ScoreClock'
	Compiling directory '/Users/brianheim/git/supercollider/testsuite/classlibrary'
	numentries = 847100 / 12777952 = 0.066
	5456 method selectors, 2342 classes
	method table size 13030072 bytes, big table size 102223616
	Number of Symbols 12367
	Byte Code Size 567090
	compiled 374 files in 0.69 seconds
compile done
localhost : setting clientID to 0.
internal : setting clientID to 0.
QObject::moveToThread: Current thread (0x7fe84cd000a0) is not the object's thread (0x7fe84d8674e0).
Cannot move to target thread (0x7fe84cd000a0)

You might be loading two sets of Qt binaries into the same process. Check that all plugins are compiled against the right Qt binaries. Export DYLD_PRINT_LIBRARIES=1 and check that only one set of binaries are being loaded.
Class tree inited in 0.03 seconds


*** Welcome to SuperCollider 3.9.0. *** For help type cmd-d.
sc3>
```

I also get this terminal output when starting the IDE:

```
objc[97514]: Class QMacAutoReleasePoolTracker is implemented in both /Users/brianheim/git/supercollider/build_qt510/Install/SuperCollider/SuperCollider.app/Contents/Frameworks/QtCore.framework/Versions/5/QtCore (0x114d7d2c8) and /usr/local/Cellar/qt/5.10.0/lib/QtCore.framework/Versions/5/QtCore (0x11bd472c8). One of the two will be used. Which one is undefined.
objc[97514]: Class QT_ROOT_LEVEL_POOL__THESE_OBJECTS_WILL_BE_RELEASED_WHEN_QAPP_GOES_OUT_OF_SCOPE is implemented in both /Users/brianheim/git/supercollider/build_qt510/Install/SuperCollider/SuperCollider.app/Contents/Frameworks/QtCore.framework/Versions/5/QtCore (0x114d7d340) and /usr/local/Cellar/qt/5.10.0/lib/QtCore.framework/Versions/5/QtCore (0x11bd47340). One of the two will be used. Which one is undefined.
objc[97514]: Class RunLoopModeTracker is implemented in both /Users/brianheim/git/supercollider/build_qt510/Install/SuperCollider/SuperCollider.app/Contents/Frameworks/QtCore.framework/Versions/5/QtCore (0x114d7d368) and /usr/local/Cellar/qt/5.10.0/lib/QtCore.framework/Versions/5/QtCore (0x11bd47368). One of the two will be used. Which one is undefined.
QObject::moveToThread: Current thread (0x7fee2840ba30) is not the object's thread (0x7fee28461c40).
Cannot move to target thread (0x7fee2840ba30)

You might be loading two sets of Qt binaries into the same process. Check that all plugins are compiled against the right Qt binaries. Export DYLD_PRINT_LIBRARIES=1 and check that only one set of binaries are being loaded.
QObject::moveToThread: Current thread (0x7fee2840ba30) is not the object's thread (0x7fee28461c40).
Cannot move to target thread (0x7fee2840ba30)

You might be loading two sets of Qt binaries into the same process. Check that all plugins are compiled against the right Qt binaries. Export DYLD_PRINT_LIBRARIES=1 and check that only one set of binaries are being loaded.
QObject::moveToThread: Current thread (0x7fee2840ba30) is not the object's thread (0x7fee28461c40).
Cannot move to target thread (0x7fee2840ba30)

You might be loading two sets of Qt binaries into the same process. Check that all plugins are compiled against the right Qt binaries. Export DYLD_PRINT_LIBRARIES=1 and check that only one set of binaries are being loaded.
QObject::connect: No such slot ScIDE::HelpBrowser::evaluateSelection(false) in /Users/brianheim/git/supercollider/editors/sc-ide/widgets/help_browser.cpp:141
```

scztt commented:
Try launching sclang with `export DYLD_PRINT_LIBRARIES=1`, it should tell you where and how libraries are getting loaded. On one hand, I don't think it's SO weird that multiple copies of these libs are getting loaded - you have an installed copy, and there's a copy in SC itself. However, the install fixup step at the end should points ALL library dependencies to the local copy, so this may be an indication that there are some pointing to the package copy and some that are not. Also, a quick and easy test would be to uninstall your local homebrew copy of QT and see if you can still launch sclang. If you can't, it's because something was pointing to that global copy.

brianlheim commented:
> Try launching sclang with export DYLD_PRINT_LIBRARIES=1, it should tell you where and how libraries are getting loaded.

Sure, here it is in a gist: https://gist.github.com/930025ed9d2efc8a2bcd9826b89614b8

> Also, a quick and easy test would be to uninstall your local homebrew copy of QT and see if you can still launch sclang. If you can't, it's because something was pointing to that global copy.

I tried this out; sclang and the IDE still launch just fine. >.>

scztt commented:
If I'm reading it right, the last lib to load before you start hitting the homebreq QT loads is [libqcocoa](https://gist.github.com/brianlheim/930025ed9d2efc8a2bcd9826b89614b8#file-export_dyld_print_results-L285). Maybe `otool -L` it? It should have similar load addresses as other QT frameworks embedded in the installed app, or e.g. the sclang executable itself.

patrickdupuis commented:
Quick question: will this PR take care of #2204 ?

brianlheim commented:
yes!

brianlheim commented:
> If I'm reading it right, the last lib to load before you start hitting the homebreq QT loads is libqcocoa. Maybe otool -L it? It should have similar load paths as other QT frameworks embedded in the installed app, or e.g. the sclang executable itself.

@scztt I tried using `otool` (didn't know about this before 😄 ) on everything in the package:

`for file in `find .`; do otool -L $file 2>/dev/null | grep qt; done`

nothing found, though... I did run it on a couple of the binaries specifically too

scztt commented:
@brianlheim I think the change I just pushed should fix. Windows may need a comparable change to it's plugin loading - note that we're got separate plugin loading cmake crap for both win and mac... this should probably be consolidated.

brianlheim commented:
@scztt yep, that works now, thanks! Things that are still problematic that I've noticed:

- lag time between evaluating code and result appearing in the post window
- Travis failure, happens for macOS only, when calling the launch script
- process for help browser is generic "QtWebEngineProcess"

I could look into any of these if you want. Second is possibly easier for me.

- merge conflict is stopping CI from running (are you OK with me merging develop into this branch to resolve this?)

I occasionally get this warning while using the help browser; a reproducer seems to be:

1. go to the "search" page
2. enter any characters in the search box
3. Ctrl-A

```
2018-02-09 20:38:01.243 QtWebEngineProcess[34940:480581] Couldn't set selectedTextBackgroundColor from default ()
```

scztt commented:
This is passing unit tests now, and I cleaned up HiDPI support significantly. I'd like to get this merged - so it can get wider testing, and because it's an ongoing burden to maintain a huge change like this out-of-mainline :) 

Regarding the remaining issues:
- The delayed post window thing @brianlheim mentioned - I have a fix for this but it's rather technical. I'd prefer to treat it as a bug and submit a separate pull request, so there can be discussion about the details of the fix. 
- The `QtWebEngineProcess` name is not trivial to fix - I wasn't able to find any clear documentation or accounts of this being done. To be honest, getting webengine packaged, installed, and launching was a multi-day CMake/QT nightmare, and I'm hesitant to fiddle with it now that it's working.
- I added a bunch of HiDPI fixes/improvements that should deal with some reported and unreported issues. Probably there will be more things to fix (since these problems are highly platform specific), but we really can't deal with those without wider testing.
- If there are any other issues related to the qt5.9 upgrade, it would be much better to deal with them separately, as bug fixes, so regular discussion and review can happen.

brianlheim commented:
> This is passing unit tests now, and I cleaned up HiDPI support significantly. I'd like to get this merged - so it can get wider testing, and because it's an ongoing burden to maintain a huge change like this out-of-mainline :)

I agree, I think it's pretty much ready for merge at this point! I can any still-existing concerns as separate issues. Mind if I give this a quick review today or tomorrow? There are a few things I saw that I have questions about.

brianlheim commented:
There are still issues with this on Windows unfortunately. I can try to help sometime this weekend. It's not being caught by AppVeyor because I haven't yet done the legwork to add a launch test to our script there.

scztt commented:
This will probably take moving some post-install commands out of a mac-specific if-block to a mac||win. I don't have easy access to a system I can test this on unfortunately, but happy to help if you're working on it. Basically, all the added cmake stuff in the diff should either be win||mac, or have a win equivalent.

brianlheim commented:
I tried building with Qt 5.5. This branch requires at least 5.7 due to use of the `QWebEngineContextMenuData` module. I haven't confirmed that it builds with 5.7 yet, going to try sometime soon.

brianlheim commented:
Update: it builds with 5.7! And the app package seems to run fine. However, cmake fails during the install phase, but I'm not entirely sure that isn't my fault. Here's a build log:

https://gist.github.com/20cf2432f2da4a1b6c7da4eb24ec730a

telephon commented:
walters@doubtfulpalace.com posted:

> 
> 1. Get info (cmd-I) on the app in the Finder
> 2. Check "Open in Low Resolution"
> 3. Your performance problems are solved
> 
> Before: 2400 sine waves (one server), very spiky CPU
> After: 9000 sine waves (one server), nice and smooth CPU
> 
> MacBook Pro (15-inch, 2016) (first touchbar model)
> High Sierra 10.13.2
> 
> Qt seems to have Retina support these days: <http://doc.qt.io/qt-5/highdpi.html>. Dunno how hard it would be to update SC to use it.

brianlheim commented:
I have this building on Windows now; I'm currently trying to clean up the installation commands for Windows and macOS.

brianlheim commented on HelpSource/Classes/WebView.schelp:
```diff
@@ -12,107 +12,296 @@ PRIVATE:: key
+prev.action 	= { webview.back };
+next.action 	= { webview.forward };
+urlBox.action 	= { webview.url = urlBox.string };
+go.action 		= { webview.url = urlBox.string };
```
tabs to align ='s here and above - this is not really style-consistent. would you mind changing it?

brianlheim commented on QtCollider/QWidgetProxy.cpp:
```diff
@@ -379,15 +379,19 @@ bool QWidgetProxy::interpretMouseWheelEvent( QObject *o, QEvent *e, QList<QVaria
   // calculate degrees: delta is in 1/8 of a degree.
-  int deg = we->delta() / 8;
-
+  // int deg = we->delta() / 8;
```
these commented lines of code can be removed

brianlheim commented on QtCollider/image.h:
```diff
@@ -184,6 +188,30 @@ class Image
+			default:
+				break;
+		}
+	}
```
this should be indented in 4 spaces

brianlheim commented on QtCollider/metatype.cpp:
```diff
@@ -124,6 +127,9 @@ MetaType *MetaType::find( PyrSlot *slot )
       }
+      else if( isKindOfSlot( slot, SC_CLASS(QCallback) ) ) {
+        return metaType<QcCallback*>();
+      }
```
NICE

brianlheim commented on QtCollider/primitives/prim_QPen.cpp:
```diff
@@ -691,10 +679,24 @@ QC_QPEN_PRIMITIVE( QPen_DrawImage, 5, PyrSlot *r, PyrSlot *a, VMGlobals *g )
+	else {
+		result = errWrongType;
+	}
+	
```
should be indented in 4-spaces

brianlheim commented on QtCollider/style/ProxyStyle.cpp:
```diff
@@ -33,7 +33,7 @@ void ProxyStyle::drawComplexControl ( ComplexControl ctrl, const QStyleOptionCom
   // FIXME: this is a workaround for the WebKit bug #104116 (or a variation on it).
   if( ctrl == QStyle::CC_ScrollBar
-      && qobject_cast<const QWebView*>(w) != 0
+      && qobject_cast<const QWebEngineView*>(w) != 0
```
Is this code needed anymore? it says workaround for WebKit

brianlheim commented on QtCollider/type_codec.hpp:
```diff
@@ -41,13 +41,14 @@
 namespace QtCollider {
 
-template <typename T> struct TypeCodec { };
+template <typename T, typename EnabledT=void> struct TypeCodec { };
```
this basically allows you to implement a fallback specialization below, right?

brianlheim commented on QtCollider/type_codec.hpp:
```diff
@@ -425,6 +438,34 @@ template <> struct TypeCodec<QVariantList>
+      QObjectT action = qobject_cast<QObjectT>(proxy->object());
+      return action;
+    } else {
+      return 0;
```
should this be `nullptr`, and then specialize the template as

```cpp
template<typename QObjectT>
struct TypeCode<QObjectT*, void>
```

which is a bit clearer in intent and clearer as a compile error if the specialization fails?

since the casts are very explicit, this might be a nice place to use `auto`, as in

```cpp
auto proxy = TypeCode<QObjectProxy*>::safeRead(slot);
```

also, a brief comment or two on this specialization couldn't hurt :)

brianlheim commented on QtCollider/widgets/QcCanvas.h:
```diff
@@ -29,8 +29,12 @@
 
-class QcCanvas : public QWidget
+//typedef QGLWidget QcCanvasBase;
+typedef QWidget QcCanvasBase;
```
I think we should prefer `using` to `typedef` in modern C++

brianlheim commented on QtCollider/widgets/QcScopeShm.cpp:
```diff
@@ -247,15 +252,18 @@ void QcScopeShm::paint1D( bool overlapped, int chanCount, int maxFrames, int fra
       min = max = frameData[0];
 
-      while( p++ < w )
+      while( (p += ratio) < w )
```
this would be better as

```cpp
for (qreal p = 0; p < w; p += ratio)
```

brianlheim replied:
oops, got this wrong, it should start at `qreal p = ratio`.

brianlheim replied:
thinking about this more, this is a bug in your change - in the previous version, `p++ < w` would execute an iteration for `p == w`. now, the loop will terminate before `p == w`. do you want that?

scztt replied:
You're right - I think the `for` construction you mentioned is the correct one.

brianlheim commented on QtCollider/widgets/QcWebView.cpp:
```diff
@@ -20,126 +20,226 @@
+          this, SLOT(onPageReload()) );
+
+  connect (page, SIGNAL(renderProcessTerminated(RenderProcessTerminationStatus, int)),
+          this, SLOT(onRenderProcessTerminated(RenderProcessTerminationStatus, int)) );
```
a more connected world :+1:

brianlheim commented on QtCollider/widgets/QcWebView.cpp:
```diff
@@ -20,126 +20,226 @@
+    size_t i = 0;
+    for (int val : data) {
+      byteData.push_back((char)val);
+}
```
`i` is unused. `}` on this line is misaligned. this is also memory-inefficient, you could call `byteData.reserve(data.size())` to get the right amount of memory immediately.

brianlheim commented on QtCollider/widgets/QcWebView.cpp:
```diff
@@ -20,126 +20,226 @@
+      byteData.push_back((char)val);
+}
+    page()->setContent(byteData, mimeType, baseUrl);
+  }
```
is there an `else` here? if not, consider adding a comment so it's clear it wasn't neglected

brianlheim commented on QtCollider/widgets/QcWebView.cpp:
```diff
@@ -20,126 +20,226 @@
+      page()->toHtml(cb->asFunctor());
+    } else {
+      page()->toHtml([](const QString&){});
+}
```
closing brace misaligned here

brianlheim commented on QtCollider/widgets/QcWebView.cpp:
```diff
@@ -20,126 +20,226 @@
+      page()->toPlainText(cb->asFunctor());
+    } else {
+      page()->toPlainText([](const QString&){});
+}
```
closing brace misaligned

brianlheim commented on QtCollider/widgets/QcWebView.cpp:
```diff
@@ -20,126 +20,226 @@
+      page()->runJavaScript(script, cb->asFunctor());
+    } else {
+      page()->runJavaScript(script, [](const QVariant&){});
+}
```
closing brace misaligned

brianlheim commented on QtCollider/widgets/QcWebView.cpp:
```diff
@@ -20,126 +20,226 @@
-  p->setDelegateReload( flag );
+  if (page()) {
+    page()->settings()->setAttribute((QWebEngineSettings::WebAttribute)attr, on);
+}
```
closing brace misaligned

brianlheim commented on QtCollider/widgets/QcWebView.cpp:
```diff
@@ -20,126 +20,226 @@
-  if( frame ) frame->evaluateJavaScript( script );
+  if (page()) {
+    page()->settings()->resetAttribute((QWebEngineSettings::WebAttribute)attr);
+}
```
closing brace misaligned

brianlheim commented on QtCollider/widgets/QcWebView.cpp:
```diff
@@ -20,126 +20,226 @@
+    QWebEngineView::findText(searchText, flags);
+  } else {
+    QWebEngineView::findText(searchText, flags, cb->asFunctor());
+}
```
misaligned brace, also consider switching the condition so that the positive branch is first

brianlheim commented on QtCollider/widgets/QcWebView.cpp:
```diff
@@ -151,25 +251,23 @@ void WebView::contextMenuEvent ( QContextMenuEvent * event )
     QMenu menu;
 
-    QPoint pos = event->pos();
+  const QWebEngineContextMenuData& contextData = page()->contextMenuData();
```
would be a lovely place for `auto`... :)

brianlheim commented on QtCollider/widgets/QcWebView.cpp:
```diff
@@ -151,25 +251,23 @@ void WebView::contextMenuEvent ( QContextMenuEvent * event )
-    menu.addAction( pageAction(QWebPage::Reload) );
+  menu.addAction( pageAction(QWebEnginePage::Back) );
+  menu.addAction( pageAction(QWebEnginePage::Forward) );
+  menu.addAction( pageAction(QWebEnginePage::Reload) );
```
this part of the file is in 4-spaces for whatever reason. consider changing it all to 2-spaces or your additions to 4-spaces

brianlheim commented on QtCollider/widgets/QcWebView.cpp:
```diff
@@ -190,7 +288,40 @@ void WebView::keyPressEvent( QKeyEvent *e )
+    return p->delegateNavigation();
+  } else {
+    return false;
+  }
```
could be simplified to a `?:` operator

brianlheim commented on QtCollider/widgets/image_painter.h:
```diff
@@ -95,6 +95,9 @@ struct ImagePainter
         QRectF rect = sourceRect;
+		qreal ratio = pixmap.devicePixelRatio();
+		rect.setWidth(rect.width() / ratio);
+		rect.setHeight(rect.height() / ratio);
```
should be spaces

brianlheim commented on SCClassLibrary/Common/GUI/Base/QWebView.sc:
```diff
@@ -1,49 +1,172 @@
-	html_ { arg htmlString;
-		this.invokeMethod( \setHtml, htmlString );
+	toHtml {
+		|cb|
```
in general SC interfaces use the argument name `action` for callbacks. given that parameter names are part of the public interface I think you should consider making it either `action` or `callback` for clarity, with a strong preference for `action`.

brianlheim commented on SCClassLibrary/Common/GUI/Base/QWebView.sc:
```diff
@@ -68,73 +188,141 @@ WebView : View {
+	back 					{ this.triggerPageAction(\back) }
+	forward					{ this.triggerPageAction(\forward) }
+	stop					{ this.triggerPageAction(\stop) }
+	reload					{ this.triggerPageAction(\reload) }
```
this is really ridiculous looking with tabs at 8 spaces; could you please reformat it to use tabs only for leading whitespace, and single spaces otherwise?

brianlheim replied:
<img width="745" alt="image" src="https://user-images.githubusercontent.com/15369640/40207238-7f8ce32a-5a01-11e8-9cc5-2b748a36e343.png">

brianlheim commented on SCClassLibrary/Common/GUI/Base/QWebView.sc:
```diff
@@ -68,73 +188,141 @@ WebView : View {
+
+
+	////////////////////////////////////////////////////////////////////////
+	// DEPRECATED
```
we need to move these into the `deprecated` folder

brianlheim commented on SCClassLibrary/Common/GUI/Base/enums.sc:
```diff
@@ -86,6 +86,88 @@ QWebFontFamily {
 }
 
+QWebAttribute {
+	classvar
```
these should be `const`, not classvar

brianlheim commented on SCClassLibrary/Common/GUI/Base/enums.sc:
```diff
@@ -86,6 +86,88 @@ QWebFontFamily {
+}
+
+QWebPageAction {
+	classvar
```
these should be `const`

brianlheim replied:
I added a FIXME for this, since it applies to more than these additions

brianlheim commented on SCClassLibrary/Common/GUI/Base/enums.sc:
```diff
@@ -86,6 +86,88 @@ QWebFontFamily {
+}
+
+QRenderProcessTerminationStatus {
+	classvar
```
should be `const` (and they're also missing accessors)

brianlheim commented on cmake_modules/MacOSXBundleInfo.plist.in:
```diff
@@ -144,7 +144,7 @@
 	<string>${PROJECT_VERSION_MAJOR}.${PROJECT_VERSION_MINOR}${PROJECT_VERSION_PATCH}</string>
 	<key>NSHumanReadableCopyright</key>
-	<string>(C) 2001-2012 James McCartney et al.</string>
+	<string>(C) 2001-2018 James McCartney et al.</string>
```
:+1:

brianlheim commented on editors/sc-ide/core/main.cpp:
```diff
@@ -82,15 +81,14 @@ int main( int argc, char *argv[] )
-
     app.setStyle( new ScIDE::Style(app.style()) );
-
+  
```
here and line 51 add whitespace for no good reason...

brianlheim commented on editors/sc-ide/widgets/code_editor/sc_editor.cpp:
```diff
@@ -1313,7 +1316,15 @@ void ScCodeEditor::findReferences()
+  
+    HelpBrowserDocklet* help = MainWindow::instance()->helpBrowserDocklet();
+    if (help) {
+        if (help->browser()->helpBrowserHasFocus()) {
```
these `if`s could be combined, same below on line 1362 (and perhaps we could do a quick refactor of these methods in a new PR)

brianlheim commented on editors/sc-ide/widgets/help_browser.cpp:
```diff
@@ -47,20 +46,28 @@
+  : WebPage(browser), mBrowser(browser)
+{
+  setDelegateNavigation(true);
+  connect( this, SIGNAL(navigationRequested(const QUrl &, QWebEnginePage::NavigationType, bool)),
```
4 spaces in this file

brianlheim commented on editors/sc-ide/widgets/help_browser.cpp:
```diff
@@ -252,13 +263,23 @@ void HelpBrowser::resetZoom()
+bool HelpBrowser::helpBrowserHasFocus() const {
+    QWidget* focused = QApplication::focusWidget();
+    
+    while (focused != NULL) {
```
either `nullptr` or just test `while (focused)`

brianlheim commented on editors/sc-ide/widgets/help_browser.cpp:
```diff
@@ -356,27 +384,33 @@ void HelpBrowser::onJsConsoleMsg(const QString &arg1, int arg2, const QString &
-    if (!hitTest.linkElement().isNull()) {
-        menu.addAction( mWebView->pageAction(QWebPage::CopyLinkToClipboard) );
+    
+    const QWebEngineContextMenuData& contextData = mWebView->page()->contextMenuData();
```
another good place for auto

brianlheim commented on editors/sc-ide/widgets/help_browser.hpp:
```diff
@@ -74,11 +76,22 @@ public slots:
+  HelpWebPage(HelpBrowser* browser);
+  
+private:
+  HelpBrowser* mBrowser;
```
4 spaces in this file

brianlheim commented on editors/sc-ide/widgets/style/style.cpp:
```diff
@@ -335,7 +335,7 @@ QRect Style::subElementRect
     }
     case QStyle::SE_TabBarTabText: {
-        const QStyleOptionTabV3 *tabOption = static_cast<const QStyleOptionTabV3*>(option);
+        const QStyleOptionTab *tabOption = static_cast<const QStyleOptionTab*>(option);
```
also good for `auto`

brianlheim commented on lang/CMakeLists.txt:
```diff
@@ -208,7 +208,7 @@ if(SC_IDE OR SC_QT)
 	get_target_property(QtCore_location Qt5::Core LOCATION)
 	message(STATUS "Found Qt: " ${QtCore_location} )
-  get_filename_component(QT_BIN_PATH ${QtCore_location} DIRECTORY CACHE)
+    get_filename_component(QT_BIN_PATH ${QtCore_location} DIRECTORY CACHE)
```
😆

brianlheim commented on QtCollider/style/ProxyStyle.cpp:
```diff
@@ -33,7 +33,7 @@ void ProxyStyle::drawComplexControl ( ComplexControl ctrl, const QStyleOptionCom
   // FIXME: this is a workaround for the WebKit bug #104116 (or a variation on it).
   if( ctrl == QStyle::CC_ScrollBar
-      && qobject_cast<const QWebView*>(w) != 0
+      && qobject_cast<const QWebEngineView*>(w) != 0
```
Still wonder if this is needed anymore?

scztt replied:
Let's get this PR in, and try removing it in a separate change so we can A-B the behavior more easily. I recall this bug being fairly subtle, and we'll need to test on all platforms.

brianlheim commented on QtCollider/widgets/QcScopeShm.cpp:
```diff
@@ -247,15 +252,18 @@ void QcScopeShm::paint1D( bool overlapped, int chanCount, int maxFrames, int fra
       float min, max;
       min = max = frameData[0];
 
-      while( p++ < w )
```
thinking about this more, this is a bug in your change - in the previous version, `p++ < w` would execute an iteration for `p == w`. now, the loop will terminate before p == w. do you want that?

scztt replied:
oops, must've typed a response and not sent. you are correct here - this is a bug, should be the for loop you mentioned elsewhere.

brianlheim replied:
ok! i'll make that change tomorrow and then we should be all set :)

brianlheim replied:
ok! i'll make that change tomorrow and then we should be all set :)

brianlheim replied:
ok! i'll make that change tomorrow and then we should be all set :)

brianlheim replied:
8 days later...

patrickdupuis commented:
I wouldn't worry too much about cosmetics. The IDE's "new look" will vary with desktop environments. It's not a noticeable difference over here, and it's much nicer looking than in the above photo.

gusano commented:
@snappizz it builds fine without hassle on arch linux and the app itself looks 100% the same : )
I can confirm that the icon is gone though..

snappizz commented:
like i said, no big deal. i suggest merging as soon as brian and scott feel ready and open some issues for the work left.

brianlheim commented:
Thanks everyone for all their hard work on this. :)
